src/pip/_internal/req/__init__.py

```diff
@@ -60,39 +70,112 @@ def install_given_reqs(
             ', '.join(to_install.keys()),
         )
 
-    installed = []
+    # pre allocate installed package names
+    installed = []  # type: List[InstallationResult]
+
+    # store install arguments
+    install_args = _InstallArgs(
+        install_options=install_options,
+        global_options=global_options,
+        root=root,
+        home=home,
+        prefix=prefix,
+        warn_script_location=warn_script_location,
+        use_user_site=use_user_site,
+        pycompile=pycompile,
+    )
 
     with indent_log():
-        for req_name, requirement in to_install.items():
-            if requirement.should_reinstall:
-                logger.info('Attempting uninstall: %s', req_name)
-                with indent_log():
-                    uninstalled_pathset = requirement.uninstall(
-                        auto_confirm=True
-                    )
-            else:
-                uninstalled_pathset = None
-
+        # get a list of packages we can install in parallel
+        if parallel_install:
+            should_parallel_reqs = [
+                (i, req) for i, req in enumerate(requirements)
+                if not req.should_reinstall and req.is_wheel
+            ]
+        else:
+            should_parallel_reqs = []
+
+        if should_parallel_reqs:
+            # prepare for parallel execution
+            should_parallel_indexes, should_parallel_values = zip(
+                *should_parallel_reqs)
+            # install packages in parallel
+            executed_parallel_reqs = map_multiprocess_ordered(
+                partial(_single_install, install_args, suppress_exception=True),
+                should_parallel_values
+            )
+            # collect back results
+            parallel_reqs_dict = dict(
+                zip(should_parallel_indexes, executed_parallel_reqs))
+        else:
+            parallel_reqs_dict = {}
```

Comment:
Should change this name (and `installed_req` below) to something like `parallel_install_results` since the values in it are not requirements.

Reply:
Sure, thing.
FYI:
I had to re-base with some new auto-black formatting for code that is unrelated see `tox auto-black` commit.